peer/pendingheap/heap.go

```diff
@@ -166,6 +169,26 @@ func (ph *pendingHeap) pushPeer(ps *peerScore) {
 	heap.Push(ph, ps)
 }
 
+// pushPeerRandom must be called in the context of a lock.
+//
+// It inserts a peer randomly into the heap among equal peer scores. This is
+// expected to be called only by Add.
+//
+// This ensures that batches of peer updates are inserted randomly throughout
+// the heap, preventing hearding behavior that may be observed during batch
+// deployments.
+func (ph *pendingHeap) pushPeerRandom(ps *peerScore) {
+	// We choose a random number between [next-len(peers), next], after
+	// incrementing ph.next
+	ph.next++
+
+	// get a random number within the range of len(peers)
+	random := ph.Rand.Float32() * float32(len(ph.peers))
+	// assign tie breaker
+	ps.last = ph.next - int(random)
```

Comment:
Will a different behavior emerge from choosing a random number between [next - len(peers), next] than just choosing any random number? I believe this does not change the probability of interposing a value between any other two numbers. @prashantv highlighted a slightly different approach in either Muttley or TChannel, where it chooses the next value and swaps it with an existing member of the heap.

Reply:
> Will a different behavior emerge from choosing a random number between [next - len(peers), next] than just choosing any random number? 

I believe it would, my thought process:
If we chose an arbitrary number outside of `len(peers)`, then all new peers added to the list will have have `last` set to some very large number. Assuming all peers have a pendingCount of 0, and we degraded to round-robin, all new peers would be added at the end of the hypothetical peer ring. This mimics the existing behaviour we're seeing since the batch insertion order is still the same. That make sense?

Regardless, 
> ...where it chooses the next value and swaps it with an existing member of the heap.

this sounds like a better strategy

Reply:
Having reasoned through this again in my head, I notice that it’s important that we use popPeer/pushPeer in Choose, pushPeerRandom in Add. The pushPeer method ensures that the peer has a higher `last` than any other peer in rotation, which is what effects round robin order, which is why we prefer to advance "next" by only one each turn. Adding "next" and swapping with an existing peer guarantees sequentiality inside the working set, whereas picking a random number in that range creates a duplicate "last" value, so the tie breaker breaks.

So yeah, that sounds like a better strategy. Let’s roll with that. My idea would have degenerated to random selection.

Reply:
yup sounds good!